### PR TITLE
Rework mode select menu

### PR DESCRIPTION
**Pull request by F1F7Y**
image

### TODO:
- [x] Add filtering
- [ ] Agree on a way of sorting modes
- [x] Allow custom modes to set their own category
- [x] Localization

### Docs:

Exposes `NSSetModeCategory( string mode, int category )` to allow mods / gamemodes to modify their category.

Categories are difend in a global enum `eModeMenuModeCategory`

### Adds these vanilla files:
- `resource/ui/menus/mode_select.menu`

### CODE REVIEW

**Comment by F1F7Y** [comment]
Marked as ready for review. Would also like to add keyboard navigation, but due to limited time that'll have to wait for next weekend.

**Review by GeckoEidechse — CHANGES_REQUESTED** [review]
Awesome work, thanks so much for looking into it <3

Playing around with it real quick and this what I noted down:

- [ ] _Marked For Death_ is teambased so it should probably be in _PvP_ and not _FFA_.
- [ ] Right now categories are sorted alphabetically which is fine but it might make sense to sort them by popularity (not sure about that one though).
- [ ] The category filter doesn't seem to work for me. Switching through the categories it still displays all of them. Not sure what's going on there ^^"

File: Northstar.Client/mod/scripts/vscripts/ui/menu_mode_select.nut
Comment (GeckoEidechse) [inline Northstar.Client/mod/scripts/vscripts/ui/menu_mode_select.nut:1]: Note to self to add this vanilla file to main

**Comment by F1F7Y** [comment]
> Awesome work, thanks so much for looking into it <3

<3

> Right now categories are sorted alphabetically which is fine but it might make sense to sort them by popularity (not sure about that one though).

With mods adding their own gamemodes this is basically impossible. Sorting alphabetically also means some modes may get more exposure

> The category filter doesn't seem to work for me. Switching through the categories it still displays all of them. Not sure what's going on there ^^"

Looks like i broke it

**Comment by F1F7Y** [comment]
Might also be worth removing the _Custom_ category as it doesnt really fit in the _PvP_, _PvE_, ... sorting methodology

**Comment by GeckoEidechse** [comment]
> Might also be worth removing the _Custom_ category as it doesnt really fit in the _PvP_, _PvE_, ... sorting methodology

Hmm, maybe. It would be kinda nice to have a distinction between _Vanilla_, _Northstar_, and _3rd party mod_ modes but I'm not sure how to combine that with sorting into _PvP_, _PvE_, etc categories without it turning out as a bunch of categories that have like 1-2 entries... ^^"

File: Northstar.Client/mod/scripts/vscripts/ui/menu_mode_select.nut [hunk 1]
```diff
@@ -1,86 +1,550 @@
...
+}
+
+void function OnCloseModesMenu()
+{
+	try
+	{
+		DeregisterButtonPressedCallback( MOUSE_WHEEL_UP , OnScrollUp )
+		DeregisterButtonPressedCallback( MOUSE_WHEEL_DOWN , OnScrollDown )
+	}
+	catch ( ex ) {}
```
Comment (uniboi) [inline Northstar.Client/mod/scripts/vscripts/ui/menu_mode_select.nut:177]: not a great solution

File: Northstar.Client/mod/scripts/vscripts/ui/menu_mode_select.nut [hunk 2]
```diff
@@ -1,86 +1,550 @@
...
+		return
+
+	var sliderButton = Hud_GetChild( file.menu , "BtnModeListSlider" )
+	var sliderPanel = Hud_GetChild( file.menu , "BtnModeListSliderPanel" )
+	var movementCapture = Hud_GetChild( file.menu , "MouseMovementCapture" )
+
+	Hud_SetFocused( sliderButton )
+
+	float minYPos = -40.0 * ( GetScreenSize()[1] / 1080.0 )
+	float maxHeight = 596.0  * ( GetScreenSize()[1] / 1080.0 )
```
Comment (uniboi) [inline Northstar.Client/mod/scripts/vscripts/ui/menu_mode_select.nut:360]:
```suggestion

    int[2] screenSize = GetScreenSize()
	float minYPos = -40.0 * ( screenSize[1] / 1080.0 )
	float maxHeight = 596.0  * ( screenSize[1] / 1080.0 )
```

File: Northstar.Client/mod/scripts/vscripts/ui/menu_mode_select.nut [hunk 3]
```diff
@@ -1,86 +1,550 @@
...
+	float maxYPos = minYPos - ( maxHeight - Hud_GetHeight( sliderPanel ) )
+	float useableSpace = ( maxHeight - Hud_GetHeight( sliderPanel ) )
+
+	float jump = minYPos - ( useableSpace / ( float( file.sortedModes.len() ) ) )
+
+	// got local from official respaw scripts, without untyped throws an error
+	local pos =	Hud_GetPos( sliderButton )[1]
+	local newPos = pos - mouseDeltaBuffer.deltaY
```
Comment (uniboi) [inline Northstar.Client/mod/scripts/vscripts/ui/menu_mode_select.nut:368]: could use `Hud_GetY` instead and type this but this code is 1 year old at this point so maybe not touch it

**Review by uniboi — CHANGES_REQUESTED** [review]
looks pretty good overall

File: Northstar.Client/mod/scripts/vscripts/ui/menu_mode_select.nut (hunk 1 above)
Comment (F1F7Y) [reply Northstar.Client/mod/scripts/vscripts/ui/menu_mode_select.nut:177]: tbh i dont even know why it needs to be in a try catch. Serverbrowser originally didnt have it and noone crashed?

Comment (uniboi) [reply Northstar.Client/mod/scripts/vscripts/ui/menu_mode_select.nut:177]: This crashes if the buttons aren't registered, but if your code isn't bad that should never happen

File: Northstar.Client/mod/scripts/vscripts/ui/menu_mode_select.nut [hunk 4]
```diff
@@ -1,86 +1,550 @@
...
+		file.scrollOffset = 0
+	}
+
+	UpdateVisibleModes()
+	UpdateListSliderPosition( file.sortedModes.len() )
+}
+
+bool function IsStringCategory( string str )
+{
+	return GetGameModeDisplayName( str ) == ""
```
[the remark this answers is not in the document]
Comment (F1F7Y) [reply Northstar.Client/mod/scripts/vscripts/ui/menu_mode_select.nut:472]: No

File: Northstar.Client/mod/resource/ui/menus/mode_select.menu [hunk 5]
```diff
@@ -0,0 +1,608 @@
...
+			tall					f0
+			visible					1
+			controlSettingsFile		"resource/ui/menus/panels/matchmaking_status.res"
+		}
+
+		MenuTitle
+		{
+			ControlName				Label
+			InheritProperties		MenuTitle
+			labelText				"#SELECT_GAME_MODE"
```
Comment (Alystrasz) [inline Northstar.Client/mod/resource/ui/menus/mode_select.menu:45]: I presume this is a vanilla entry?

File: Northstar.Client/mod/resource/ui/menus/panels/mode_select_button.res [hunk 6]
```diff
@@ -0,0 +1,31 @@
+resource/ui/menus/panels/mode_select_button.res
+{
+	BtnMode
+	{
+		ControlName 			RuiButton
+		InheritProperties 		RuiSmallButton
+		classname 				ModButton
+		labelText				"please show up"
```
Comment (Alystrasz) [inline Northstar.Client/mod/resource/ui/menus/panels/mode_select_button.res:8]: I didn't see this text while testing :)

File: Northstar.Client/mod/scripts/vscripts/ui/menu_mode_select.nut [hunk 7]
```diff
@@ -1,86 +1,552 @@
...
+	UNKNOWN = 0,
+	PVPVE   = 1,
+	PVE     = 2,
+	PVP     = 3,
+	FFA     = 4,
+	TITAN   = 5,
+	OTHER   = 6,
+	CUSTOM  = 7
+
+	SIZE
```
Comment (Alystrasz) [inline Northstar.Client/mod/scripts/vscripts/ui/menu_mode_select.nut:16]: What's the meaning of this?

File: Northstar.Client/mod/scripts/vscripts/ui/menu_mode_select.nut [hunk 8]
```diff
@@ -1,86 +1,552 @@
...
+	table<string,int> categoryOverrides
+
+	// List of all modes we know
+	array<ListEntry_t> modes
+
+	// Sorted list of modes we want to show with categories included
+	array<string> sortedModes
 } file
 
 const int MODES_PER_PAGE = 15
```
Comment (Alystrasz) [inline Northstar.Client/mod/scripts/vscripts/ui/menu_mode_select.nut:58]: Since you're not using pages, why keeping this?

**Review by Alystrasz — APPROVED** [review]
Code mostly looks good.
I confirm new menu and gamemodes filtering work as expected.

File: Northstar.Client/mod/resource/ui/menus/mode_select.menu (hunk 5 above)
Comment (F1F7Y) [reply Northstar.Client/mod/resource/ui/menus/mode_select.menu:45]: This is the giant menu title in the top-right

File: Northstar.Client/mod/resource/ui/menus/panels/mode_select_button.res (hunk 6 above)
Comment (F1F7Y) [reply Northstar.Client/mod/resource/ui/menus/panels/mode_select_button.res:8]: `RuiButton` has 2 text fields, the old `labelText` and then the new rui text. We hide `labelText` as we only use the rui one. This is here for testing as buttons are hard to spot without text.

File: Northstar.Client/mod/scripts/vscripts/ui/menu_mode_select.nut (hunk 8 above)
Comment (F1F7Y) [reply Northstar.Client/mod/scripts/vscripts/ui/menu_mode_select.nut:58]: The scroll bar logic uses this in multiple places. Imo it's better to keep this as a constant for readability.

File: Northstar.Client/mod/scripts/vscripts/ui/menu_mode_select.nut (hunk 7 above)
Comment (F1F7Y) [reply Northstar.Client/mod/scripts/vscripts/ui/menu_mode_select.nut:16]: `SIZE` will always be 1 more than the last entry in the enum. This allows the enum to be modified without having to adapt [other code](https://github.com/R2Northstar/NorthstarMods/pull/624/files#diff-0acf05adff8e31dbe3df62eba236d9001887b39f34abd248ed449a26ea59322cR87-R94)

**Review by uniboi — APPROVED** [review]
Code looks good. Modes appear now categorized in the menu.

**Comment by GeckoEidechse** [comment]
@F1F7Y PR has merge conflict atm. If you fix it, I can test ^^

**Comment by F1F7Y** [comment]
i am going to scream

**Comment by ASpoonPlaysGames** [comment]
> [ ] Agree on a way of sorting modes

Has this been done in any capacity?

> Hmm, maybe. It would be kinda nice to have a distinction between _Vanilla_, _Northstar_, and _3rd party mod_ modes but I'm not sure how to combine that with sorting into _PvP_, _PvE_, etc categories without it turning out as a bunch of categories that have like 1-2 entries... ^^"

This seems more like gamemode "tags" to me, could be added in a future PR with sorting and stuff maybe?

**Comment by uniboi** [comment]
> > [ ] Agree on a way of sorting modes

>Has this been done in any capacity?

I think the current way is fine tbh

**Comment by F1F7Y** [comment]
re-requested gecko as i cant push to main

**Comment by F1F7Y** [comment]
@GeckoEidechse bump

**Comment by GeckoEidechse** [comment]
Merge conflicts atm ;-;

**Comment by GeckoEidechse** [comment]
bump

**Comment by GeckoEidechse** [comment]
Merge conflicts still

**Comment by GeckoEidechse** [comment]
Bump cause merge conflicts

**Comment by GeckoEidechse** [comment]
Removed "READY TO MERGE" lable until merge conflicts are resolved

**Comment by GeckoEidechse** [comment]
Bump

File: Northstar.Client/mod/resource/ui/menus/panels/mode_select_button.res
Comment (GeckoEidechse) [inline Northstar.Client/mod/resource/ui/menus/panels/mode_select_button.res:1]: This file is entirely custom right? At least I couldn't find it in the frontend VPK... \^\^

File: Northstar.Client/mod/resource/ui/menus/mode_select.menu
Comment (GeckoEidechse) [inline Northstar.Client/mod/resource/ui/menus/mode_select.menu:1]: Added vanilla file via bf5f6b02e8ad74ece0aa43d21a6cec88db6641ed

File: Northstar.Client/mod/resource/ui/menus/panels/mode_select_button.res
Comment (F1F7Y) [reply Northstar.Client/mod/resource/ui/menus/panels/mode_select_button.res:1]: Should be yeah

**Review by GeckoEidechse — COMMENTED** [review]
Changes confirmed working in testing.

image

List is nicer than the previous vanilla one. As for the exact ordering I'd say we decide on that post-merge ^^

(Not marking as approved until I got the file diff figured out mentioned in the comment above. After that it's ready to merge :D)

File: Northstar.Client/mod/resource/ui/menus/mode_select.menu
[the remark this answers is not in the document]
Comment (Alystrasz) [reply Northstar.Client/mod/resource/ui/menus/mode_select.menu:1]: Confirmed using VSCode:

##### Main branch:
mainmain

##### F1F7Y's branch:
main-F1F7Y

[the remark this answers is not in the document]
Comment (GeckoEidechse) [reply Northstar.Client/mod/resource/ui/menus/mode_select.menu:1]: Thanks for confirming. This is what I get for committing a CRLF file from Linux... (/.-)

Fixed in e2931190f9b37fc4ec171cd348d85eb96ed591a3

**Review by GeckoEidechse — APPROVED** [review]
Fixed line-endings on main. This should be good to merge now ^^

**Comment by GeckoEidechse** [comment]
Welp so somehow I missed that it kinda just marks all gamemodes as locked, fuck

I really shouldn't test PRs while exhausted x_x

**Pull request merged**